### ISSUE TITLE
How to make Formie work with Craft Cloud static caching?

### ISSUE THREAD

**Issue by bleepsandblops** [verbb/formie#2179]
### Describe the bug

Hello,

It would be great to get a simple guide on how to make Formie work with ajax submissions specifically with Craft Cloud and the way it handles caching.

I've tried a couple of things, and none of them seem to work.

1. I tried adding `{{ craft.formie.renderJs(true) }}` in the head but nothing gets loaded if I look at the source. The plan was to then have Formie available within inline js or within my JS components. (If it's helpful, I'm also using Sprig and it gets automatically loaded fine in the footer because there's a sprig tag on the page).
2. I tried including it within a js component with `import { Formie } from '../../../vendor/verbb/formie/src/web/assets/frontend/src/js/formie-lib';` but it's not working because the path isn't correct.
3. Tried adding `{{ craft.formie.renderFormJs('contact') }}` in the footer to trigger the JS load on each page, but then i'm getting a cache MISS 

Thank you!

### Steps to reproduce

1. See above

### Form settings

- No
- Submission Method: Ajax 
- Client-side Validation: Yes 
- Custom Form Templates: No

### Craft CMS version

5.5.5

### Plugin version

5.0.14

### Multi-site?

No

**Comment by engram-design** [comment]
Formie should be working with Craft Cloud, so long as you follow the [caching](https://verbb.io/craft-plugins/formie/docs/template-guides/cached-forms) guide.

I don't see any reason why `renderJs` (inline) wouldn't work, it's my understanding that should be fine.

Using option 2 isn't going to really work, as the paths are going to be messed up, based on the filesystem in use.

I'll test out Craft Cloud shortly. Out of interest, does a `craft.formie.renderForm('contact')` call work (which loads resources unless you have a Form Template set for the form)?

**Comment by bleepsandblops** [comment]
> I'll test out Craft Cloud shortly. Out of interest, does a `craft.formie.renderForm('contact')` call work (which loads resources unless you have a Form Template set for the form)?

It 'works' but then the page it's on never caches, always MISS cache status.

**Comment by bleepsandblops** [comment]
@engram-design could this be the issue re `renderJs` not working (as I can see you're using `getPublishedUrl`)

`Any static assets your plugin provides to the control panel or front-end must be encapsulated in an [Asset Bundle](https://craftcms.com/docs/5.x/extend/asset-bundles.html), and its sourcePath must begin with your plugin’s predefined alias. To register an asset bundle, call Craft::$app->getView()->registerAssetBundle($myBundle) from a controller or template. Publishing one-off or ad-hoc assets at runtime is not supported on Cloud.`

https://craftcms.com/knowledge-base/cloud-plugin-development#asset-bundles

https://github.com/wrav/oembed/issues/158

**Comment by engram-design** [comment]
Nah, we're ensuring everything is using an asset bundle. Do you have the [asyncCsrfInputs](https://craftcms.com/docs/5.x/reference/config/general.html#asynccsrfinputs) enabled?

**Comment by bleepsandblops** [comment]
Ah-ha! So basically after enabling asyncCsrfInputs and going back to a super basic setup of having just `{{ craft.formie.renderForm(form) }}`, it just works... the page is cached and the forms work nicely, and i can see the token refresh call in the XHR network tab.

Thanks a lot!

**Comment by engram-design** [comment]
I should certainly add a Craft Cloud section to the docs at least, so I appreciate the suggestion! There's still one or two things I want to test and properly confirm, but from what others tell me, it's _mostly_ working.

**Comment by jamesmacwhite** [comment]
Turns out this isn't correct, but PHPStorm is wrongly saying it is default true. That's fun!

**Comment by jamesmacwhite** [comment]
For full context, specifically with Craft Cloud, having `asyncCsrfInputs` set to true is pretty much essential. The reason being with Craft Cloud and static caching enabled, the async CSRF input setting performs a .fetch() XHR call to get a fresh CSRF token and then does a document.querySelectorAll() to find all the CSRF input fields on a page and injects a fresh one via JavaScript/AJAX.

You'll find this JS injected into your web application:

```js
  (function() {
    fetch('https://example.craft.site/actions/users/session-info', {
      headers: {
        'Accept': 'application/json',
      }
    }).then(response => response.json())
      .then(data => {
        document.querySelectorAll('craft-csrf-input')
          .forEach(element => {
            const input = document.createElement('input');
            input.type = 'hidden';
            input.name = data.csrfTokenName;
            input.value = data.csrfTokenValue;
            element.replaceWith(input);
          });
      });
  })();
```

The static caching of Craft Cloud is smart enough to know when you include a non-async CSRF token in your HTML, it turns off any caching entirely, because it knows it isn't safe to cache that request. This means you can't do a typical pattern of stashing a CSRF token value in your HTML somewhere to retrieve later, you must do async CSRF everywhere to be compatible with it.

You can of course send no cache headers, but that will happen automatically anyway and the performance hit of not having static caching enabled isn't ideal.

**Comment by joepagan** [comment]
This is not just a Craft Cloud issue, any website with any static caching service has the same issue.

Where `asyncCsrfInputs` is `false`, this will force pages with a form on it to return with a `no-cache` header, meaning it bypasses the caching layer.

This is an issue where `asyncCsrfInputs` came into play - with craft 4 (think it came in from v4.9) and craft 5 (not sure)

The issue here is that formie's `renderForm` function does not include an `input` field for the csrf input, instead the new custom element `craft-csrf-input` is used and the craftcms script @jamesmacwhite highlighted will replace this element with an input, but the script from formie's docs does not account for the craftcms script not having completed and injecting an input element into the form existing at the point in time formie's script runs. I've implemented this in which overcomes it (feel free to use/modify @engram-design ):

composer.json
```
"craftcms/cms": "4.16.13",
"verbb/formie": "2.2.5",
```

template.twig
```twig
{% set form = block.form.one ?? null %}
{{ craft.formie.renderForm(form) }}
{% include '_scripts/csrf' with { form: form } %}
```

csrf.twig
```js
{% if form %}
  {% js %}
    // Wait until Formie has been loaded and initialised
    document.addEventListener('onFormieInit', (event) => {
      console.log('Formie initialised');
      // Fetch the Form Factory once it's been loaded
      let Formie = event.detail.formie;

      let $form = document.querySelector('#{{ form.formId }}');

      // Find the CSRF token hidden input, so we can replace it
      let $csrfInput = $form.querySelector('input[name="CRAFT_CSRF_TOKEN"]');

       // Check CSRF input exists, if not, check every 300ms up to 10 times until it does
       async function waitForCsrfInput() {
         if ($csrfInput) {
           return $csrfInput;
         }
         
         console.log('CSRF input not found, checking again');
         
         for (let i = 0; i < 10; i++) {
           $csrfInput = $form.querySelector('input[name="CRAFT_CSRF_TOKEN"]');
           if ($csrfInput) {
             console.log('CSRF input found after '+ (i + 1) +' attempts ('+(300*(i+1))+'ms) for form {{ form.formId }}');
             return $csrfInput;
           }
           
           if (i < 9) { // Don't wait after the last attempt
             await new Promise(resolve => setTimeout(resolve, 300));
           }
         }
         
         const msg = 'CSRF input not found after 10 attempts';
         console.error(msg);
         throw new Error(msg);
       }

       waitForCsrfInput().then(() => {
         console.log('CSRF input found, refreshing form cache');
         // Refresh the necessary bits that are statically cached (CSRF inputs, captchas, etc)
         Formie.refreshForCache('{{ form.formId }}');
       }).catch((error) => {
         console.error('Failed to find CSRF input:', error);
       });
    });
  {% endjs %}
{% else %}
  {% js %}
    const msg = 'Form not passed to CSRF script';
    console.error(msg);
    throw new Error(msg);
  {% endjs %}
{% endif %}
```

**Comment by engram-design** [comment]
Appreciate the input @joepagan, absolutely on my list to revisit this!

**Comment by engram-design** [comment]
Updated in [4.0.0-beta.1](https://github.com/verbb/formie/releases/tag/4.0.0-beta.1).
